src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
     log_println(3, "Successfully returned from run_test() routine");
   } else {
     log_println(3,
-                "Child %d returned non-zero (%d) from "
-                "run_test() results some test failed!",
+                "Child %d returned non-zero (%d) from run_test() results some "
+                "test failed!",
                 getpid(), retcode);
-    child_sig(0);
   }
-  SSL_free(ctl.ssl);
-  close(ctl.socket);
+  close_connection(&ctl);
 #if USE_WEB100
   web100_detach(agent);
 #elif USE_WEB10G
   estats_nl_client_destroy(&agent);
 #endif
-  // log_free(); // Don't free the log we use it all the time
-  // log_println()
-  // Also makes valgrind angry
 
   if (cputime && workerThreadId) {
     cputimeworkerLoop = 0;
     pthread_join(workerThreadId, NULL);
   }
-
-  // At this point the tests have been run and we need to clean up
-  // and handle and child processes that might still be lying
-  // around.  If we don't we get zombies. The pkt-pair handling
-  // created 2 childern and we need to get rid of them. To know
-  // what PIDs to look for on 1/19/10 the run_test() routine was
-  // modified to pass around these values. The values are set in
-  // the proper routine in the testoptions.c file.
-  // Then add call to child_sig() routine and pass in these PID's
-  // so we handle each child in sequence
-
-  log_println(6, "remove pkt-pair children c2s=%d, s2c=%d", testopt.child1,
-              testopt.child2);
-  if (testopt.child1 != 0) child_sig(testopt.child1);
-  if (testopt.child2 != 0) child_sig(testopt.child2);
   exit(0);
 }
 
+int max(int a, int b) { return (a > b) ? a : b; }
+
 /**
- * The server's main loop.  This is the function that, once all arguments are
- * processed and the server environment has been set up, will keep waiting for
- * new connections and then forking off children to handle those connections.
+ * Waits for a new client to connect, for one of the child processes to die, or
+ * (optionally) until a timeout occurs.  If the select() is woken up by a new
+ * client connecting, then this routine will allocate and initialize a new
+ * ndtchild struct, and fork off the child process.
+ * @param listenfd The file descriptor which listens to the network
+ * @param signalfd The pipe which will be written to upon receipt of SIGCHILD.
+ * @param waitforever True if the client should wait forever, false otherwise
+ * @param ssl_context The ssl_context for any new connections
+ * @returns NULL or a newly initialized ndtchild struct. The calling function
+ *          owns the struct.
  */
-void NDT_server_main_loop(SSL_CTX *ssl_context, int listenfd) {
+struct ndtchild *wait_for_new_child(int listenfd, int signalfd, int waitforever,
```

Comment:
Optional, arguably out of scope: I think we're touching almost every `ndtchild` with this pull request. Can we `typedef` it so that it's just `ndtchild` instead of `struct ndtchild`?